File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/TestBinder.java
package com.microsoft.signalr;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TestBinder implements InvocationBinder {
    private TypeAndClass[] paramTypes = null;
    private TypeAndClass returnType = null;

    public TestBinder(HubMessage expectedMessage) {
        if (expectedMessage == null) {
            return;
        }

        switch (expectedMessage.getMessageType()) {
            case STREAM_INVOCATION:
                ArrayList<TypeAndClass> streamTypes = new ArrayList<>();
                for (Object obj : ((StreamInvocationMessage) expectedMessage).getArguments()) {
                    Type type = getType(obj.getClass());
                    streamTypes.add(new TypeAndClass(type, obj.getClass()));
                }
                paramTypes = streamTypes.toArray(new TypeAndClass[streamTypes.size()]);
                break;
            case INVOCATION:
                ArrayList<TypeAndClass> types = new ArrayList<>();
                for (Object obj : ((InvocationMessage) expectedMessage).getArguments()) {
                    Type type = getType(obj.getClass());
                    types.add(new TypeAndClass(type, obj.getClass()));
                }
                paramTypes = types.toArray(new TypeAndClass[types.size()]);
                break;
            case STREAM_ITEM:
                break;
            case COMPLETION:
                Object obj = ((CompletionMessage)expectedMessage).getResult().getClass();
                returnType = new TypeAndClass(getType(((CompletionMessage)expectedMessage).getResult().getClass()), 
                    ((CompletionMessage)expectedMessage).getResult().getClass());
                break;
            default:
                break;
        }
    }
    
    private <T> Type getType(Class<T> param) {
        return (new TypeReference<T>() {}).getType();
    }

    @Override
    public TypeAndClass getReturnType(String invocationId) {
        return returnType;
    }

    @Override
    public List<TypeAndClass> getParameterTypes(String methodName) {
        if (paramTypes == null) {
            return new ArrayList<>();
        }
        return new ArrayList<TypeAndClass>(Arrays.asList(paramTypes));
    }
}

File: src/SignalR/perf/Microbenchmarks/TestBinder.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using Microsoft.AspNetCore.SignalR.Protocol;

namespace Microsoft.AspNetCore.SignalR.Microbenchmarks;

public class TestBinder : IInvocationBinder
{
    private readonly Type[] _paramTypes;
    private readonly Type _returnType;

    public TestBinder(HubMessage expectedMessage)
    {
        switch (expectedMessage)
        {
            case StreamInvocationMessage i:
                _paramTypes = i.Arguments?.Select(a => a?.GetType() ?? typeof(object))?.ToArray();
                break;
            case InvocationMessage i:
                _paramTypes = i.Arguments?.Select(a => a?.GetType() ?? typeof(object))?.ToArray();
                break;
            case StreamItemMessage s:
                _returnType = s.Item?.GetType() ?? typeof(object);
                break;
            case CompletionMessage c:
                _returnType = c.Result?.GetType() ?? typeof(object);
                break;
        }
    }

    public TestBinder() : this(null, null) { }
    public TestBinder(Type[] paramTypes) : this(paramTypes, null) { }
    public TestBinder(Type returnType) : this(null, returnType) { }
    public TestBinder(Type[] paramTypes, Type returnType)
    {
        _paramTypes = paramTypes;
        _returnType = returnType;
    }

    public IReadOnlyList<Type> GetParameterTypes(string methodName)
    {
        if (_paramTypes != null)
        {
            return _paramTypes;
        }
        throw new InvalidOperationException("Unexpected binder call");
    }

    public Type GetReturnType(string invocationId)
    {
        if (_returnType != null)
        {
            return _returnType;
        }
        throw new InvalidOperationException("Unexpected binder call");
    }

    public Type GetStreamItemType(string streamId)
    {
        throw new NotImplementedException();
    }
}
